[BEAM-11997] Changed RedisIO implementation to SDF

MiguelAnzoWizeline opened the pull request:
Changed RedisIO implementation to SDF

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/badge/icon?subject=V1+Streaming">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon?subject=V1+Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/badge/icon?subject=V2+Streaming">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon?subject=Java+8">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon?subject=Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon?subject=Portable+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/badge/icon?subject=Portable">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon?subject=Structured+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon?subject=ValCont">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/badge/icon?subject=V1+Java11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon?subject=3.6">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon?subject=3.7">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon?subject=3.8">
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon?subject=Tests">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon?subject=Lint">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon?subject=Docker">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon?subject=Docs">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

MiguelAnzoWizeline commented:
Run Java PreCommit

MiguelAnzoWizeline commented:
Run Java PreCommit

MiguelAnzoWizeline commented:
R: @chamikaramj

chamikaramj commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +336,88 @@ public void setup() {
-          finished = true;
+      }
+      if (bundle.size() > 0) {
+        List<KV<String, String>> kvs = fetchAndFlush(bundle);
```
Probably loading the whole bundle to memory here could result in OOMs.

MiguelAnzoWizeline replied:
Updated

chamikaramj commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +336,88 @@ public void setup() {
+        if (newStart + batchSize <= end) {
+          offsetList.add(new OffsetRange(newStart, newStart + batchSize));
+          newEnd = newStart + batchSize;
+          newStart = newStart + batchSize + 1;
```
Seems like we are loosing a record position here. If the previous split ends at (newStart + batchSize) next split should start at that. Also please add unit tests that would cover this.

MiguelAnzoWizeline replied:
Thanks, found the problem and updated, updated the tests for different cases of data size and batch size

chamikaramj commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +336,88 @@ public void setup() {
-        List<String> keys = scanResult.getResult();
-        for (String k : keys) {
-          c.output(k);
+    public void processElement(ProcessContext c, RestrictionTracker<OffsetRange, Long> tracker) {
```
Please check other source implementations and add similar tests related to splitting (and reading if anything is missing).
Here are the tests for TextSource for example (even thought it's not an SDF).
https://github.com/apache/beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOReadTest.java

chamikaramj commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +336,88 @@ public void setup() {
+      while (true) {
+        if (newStart + batchSize <= end) {
+          offsetList.add(new OffsetRange(newStart, newStart + batchSize));
+          newEnd = newStart + batchSize;
```
Probably simpler to just set correct newStart and newEnd values for the next batch and use that here.

MiguelAnzoWizeline replied:
Updated and removed newEnd to keep simpler

chamikaramj reviewed (commented):
Thanks.

MiguelAnzoWizeline commented:
Hi @lukecwik 
I'm working on the changes mentioned, I have one question, if I decide to use the ByteKeyRangeTracker instead of the new RedisCursorRangeTracker that I was implementing, how would updating the range work? would I need to update the range to [nextCursor, end] after every loop? If so, would that be done by creating a new instance of the tracker with the updated range in every loop? and in that case would the end restriction is going to always be 0?

lukecwik commented:
Your loop `ProcessElement` loop would be like:
```
byteKey = tracker.getRestriction().getStart();
while (tryClaim(byteKey)) {
  cursor = convertToRedisCursor(byteKey);
  values, nextCursor = jedis.scan(cursor);
  byteKey = convertToByteKey(nextCursor);
  output values
}
return STOP;
```

You could also create a `RedisCursorRangeTracker` that is a thin wrapper over the `ByteKeyRangeTracker` and effectively do conversions only by internalizing the `convertToRedisCursor` and `convertToByteKey` methods.

MiguelAnzoWizeline commented:
> Your loop `ProcessElement` loop would be like:
> 
> ```
> byteKey = tracker.getRestriction().getStart();
> while (tryClaim(byteKey)) {
>   cursor = convertToRedisCursor(byteKey);
>   values, nextCursor = jedis.scan(cursor);
>   byteKey = convertToByteKey(nextCursor);
>   output values
> }
> return STOP;
> ```
> 
> You could also create a `RedisCursorRangeTracker` that is a thin wrapper over the `ByteKeyRangeTracker` and effectively do conversions only by internalizing the `convertToRedisCursor` and `convertToByteKey` methods.

Hi @lukecwik 
So, in the case of making a wrapper over `ByteKeyRangeTracker` we would have the `RedisCursorRangeTracker` and that would have an **internal** `ByteKeyRangeTracker`. When calling the `RedisCursorRangeTracker` `tryClaim` method this would make the conversion from `RedisCursor` to `ByteKey` and call the internal tracker `tryClaim` method, it would also be in this method were it would be appropriate to handle the special case when both start and end positions are the `0` cursor, and were we could update the range to be `[previousEndCursor, newCursor]` to avoid the end cursor to always be `0` since that would always fail the `tryClaim`, would that be correct?

lukecwik commented:
Sounds about right. It might be easier if you make `START` as a static constant so you can differentiate it from `END`.

MiguelAnzoWizeline commented:
@lukecwik Hi,
I was testing the implementation that we considered in the previous message, using an internal ByteKeyRangeTracker gave me some problems with keeping the range, lastAttemptedKey and lastClaimedKey updated correctly. 
The implementation with the new changes I commited works by having a ByteKey in RedisCursor along with the String cursor representation, and RedisCursorRangeTracker copies the functionality  of ByteKeyRangeTracker (with the added special conditions of the Redis implementation) using the ByteKey representation of the cursor to keep track of the range, this works well for all the read test and seems a cleaner approach.
Still missing the unit test for the new code. I wanted to see first if the changes looks correct to you.

MiguelAnzoWizeline commented:
Run Java PreCommit

MiguelAnzoWizeline commented:
@lukecwik Fixed some style issues and warnings, waiting on the review to check if my current implementation is correct.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java:
```diff
@@ -0,0 +1,149 @@
+
+  @VisibleForTesting
+  static long shiftBits(long a, int nBits) {
+    long b = 0;
```
Can you simplify this with Long.reverse?

It will also help with readability.

pabloem replied:
+1 - please use Long.reverse(a) , unless you need the `nBits` argument?

pabloem replied:
you can do the following right?

```
nBits = getTablePow(cursor.getDbSize());
Long.reverse(cursorLong) >> (64 - nBits)
```

pabloem replied:
can you please use Long.reverse?

benWize replied:
I tried changing to    ` long reversed = Long.reverse(cursorLong) >> (64 - nBits);` but the test https://github.com/apache/beam/blob/4adce5979bf7866efa7041f8399703fe31b6053a/sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java#L397` failed with:
`expected:<[1885267]> but was:<[-211885]>`

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java:
```diff
@@ -0,0 +1,149 @@
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class RedisCursor implements Comparable<RedisCursor>, Serializable {
```
It makes more sense to treat this as a simple container class that contains
```
String cursor
long dbSize
boolean isStart
```

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursorRangeTracker.java:
```diff
@@ -0,0 +1,239 @@
+  /* An empty range which contains no keys. */
+  @VisibleForTesting
+  static final RedisCursorRange NO_KEYS =
+      RedisCursorRange.of(RedisCursor.ZERO_CURSOR, RedisCursor.ZERO_CURSOR);
```
Isn't `[0, 0)` also equivalent to the all keys range?

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursorRangeTracker.java:
```diff
@@ -0,0 +1,239 @@
+public class RedisCursorRangeTracker extends RestrictionTracker<RedisCursorRange, RedisCursor>
+    implements RestrictionTracker.HasProgress {
+
+  private RedisCursorRange range;
```
I think you can you use a ByteKeyRangeTracker as a delegate object and perform RedisCursor -> ByteKey conversions and ByteKey -> RedisCursor conversions here if you had a better way to represent `0` as start and `0` as end.

`0 && isStart -> ByteKey.of(0x00)`
`0 && !isStart -> ByteKey.EMPTY`

All the other mappings will fall in between.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +373,47 @@ public void setup() {
-      for (BoundedWindow w : kvs.keySet()) {
-        for (KV<String, String> kv : kvs.get(w)) {
-          context.output(kv, w.maxTimestamp(), w);
+        if (RedisCursor.ZERO_CURSOR.getCursor().equals(scanResult.getCursor())) {
+          break;
         }
```
I would expect the next tryClaim to fail and you shouldn't have to have this exit condition here.

You want the RedisCursor to differentiate `0` as the start or end by embedding a boolean like `isStart`.

Then in the comparison `isStart` compares before everything and `0 && !isStart` compares after everything.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +373,47 @@ public void setup() {
-          c.output(k);
+      while (tracker.tryClaim(cursor)) {
+        ScanResult<String> scanResult = jedis.scan(cursor.getCursor(), scanParams);
+        List<String> keys = new ArrayList<>();
+        for (String k : scanResult.getResult()) {
+          keys.add(k);
         }
```
Why add them to another list when `scanResult.getResult()` is already a list?

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +373,47 @@ public void setup() {
-          if (results.get(i) != null) {
-            kvs.put(w, KV.of(keys[i], results.get(i)));
-          }
+    private List<KV<String, String>> fetchAndFlush(List<String> bundle) {
```
Can you move this code into the `@ProcessElement` method so we can output each key and value as it appears instead of creating another list of all the results?

(Alternatively pass in the `ProcessContext` and output directly from within this method)

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -122,6 +126,38 @@ public static Read read() {
         .build();
   }
 
+  public static ByteKey cursorToByteKey(String cursor, long dbSize) throws IOException {
+    long cursorLong = Long.parseLong(cursor);
+    long reversed = shiftBits(cursorLong, dbSize);
+    BigEndianLongCoder coder = BigEndianLongCoder.of();
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+    coder.encode(reversed, os);
+    byte[] byteArray = os.toByteArray();
+    return ByteKey.copyFrom(byteArray);
+  }
+
+  public static String byteKeyToString(ByteKey byteKeyStart) {
+    ByteBuffer bb = ByteBuffer.wrap(byteKeyStart.getBytes());
+    if (bb.capacity() < 8) {
+      int rem = 8 - bb.capacity();
+      byte[] padding = new byte[rem];
+      bb = ByteBuffer.allocate(8).put(padding).put(bb.array());
+      bb.position(0);
+    }
+    long l = bb.getLong();
+    return Long.toString(l);
+  }
+
+  public static long shiftBits(long a, long dbSize) {
+    long b = 0;
+    for (long i = 0; i < dbSize; ++i) {
+      b <<= 1;
+      b |= (a & 1);
+      a >>= 1;
+    }
+    return b;
+  }
```
Duplicate code?

lukecwik commented:
The issues you ran into with the ByteKeyRangeTracker exist because of issues with differentiating `0` as start and end.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,94 +373,47 @@ public void setup() {
-      super(connectionConfiguration);
+    @GetInitialRestriction
+    public RedisCursorRange getInitialRestriction() {
+      long dbSize = jedis.dbSize();
```
This needs to be the table size and not the dbsize.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java:
```diff
@@ -0,0 +1,144 @@
+  @Override
+  public int compareTo(@Nonnull RedisCursor other) {
+    checkNotNull(other, "other");
+    return Long.compare(Long.parseLong(cursor), Long.parseLong(other.cursor));
```
You need to differentiate start/end

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java:
```diff
@@ -0,0 +1,144 @@
+    } catch (IOException e) {
+      throw new IllegalArgumentException("invalid redis cursor " + cursor);
+    }
+    byte[] byteArray = os.toByteArray();
+    return ByteKey.copyFrom(byteArray);
+  }
+
+  @VisibleForTesting
+  static long shiftBits(long a, int nBits) {
+    long b = 0;
+    for (int i = 0; i < nBits; ++i) {
+      b <<= 1;
+      b |= (a & 1);
+      a >>= 1;
+    }
+    return b;
+  }
+
+  @VisibleForTesting
+  static int getTablePow(long nKeys) {
+    return 64 - Long.numberOfLeadingZeros(nKeys - 1);
+  }
+
+  @VisibleForTesting
+  static RedisCursor byteKeyToRedisCursor(ByteKey byteKeyStart, long nKeys, boolean isStart) {
+    if (byteKeyStart.isEmpty() || byteKeyStart.equals(ZERO_KEY)) {
+      return RedisCursor.of("0", nKeys, isStart);
+    }
+    int nBits = getTablePow(nKeys);
+    ByteBuffer bb = ByteBuffer.wrap(byteKeyStart.getBytes());
+    if (bb.capacity() < nBits) {
+      int rem = nBits - bb.capacity();
+      byte[] padding = new byte[rem];
+      bb = ByteBuffer.allocate(nBits).put(padding).put(bb.array());
+      bb.position(0);
+    }
+    long l = bb.getLong();
+    l = shiftBits(l, nBits);
+    return RedisCursor.of(Long.toString(l), nKeys, isStart);
+  }
```
It makes sense to have these methods within the RedisCursorRangeTracker since that is where you need to perform the conversions.

lukecwik commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursorRangeTracker.java:
```diff
@@ -0,0 +1,230 @@
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class RedisCursorRangeTracker extends RestrictionTracker<RedisCursorRange, RedisCursor>
+    implements RestrictionTracker.HasProgress {
```
Where is the ByteKeyRangeTracker?

lukecwik reviewed (changes requested):
Did you miss pushing up more of the PR?

MiguelAnzoWizeline commented:
Updated RedisCursor compareTo to differentiate start and end, moved util functions to RedisCursorRangeTrakcer.
About the internal ByteKeyRangeTracker, I removed it and moved its logic to the RedisCursorRangeTracker as having the internal ByteKeyRangeTracker presented some problems, mainly that the trySplit method would produce ranges with a starting key that was not a valid cursor returned by redis (because of its next function adding 1 to the current ByteKey) this caused problems on the test case with a large number of keys. 
RedisCursorRangeTracker maintains the same logic used on ByteKeyRangeTracker with the exception of a modified trySplit method.

lukecwik commented:
> Updated RedisCursor compareTo to differentiate start and end, moved util functions to RedisCursorRangeTrakcer. About the internal ByteKeyRangeTracker, I removed it and moved its logic to the RedisCursorRangeTracker as having the internal ByteKeyRangeTracker presented some problems, mainly that the trySplit method would produce ranges with a starting key that was not a valid cursor returned by redis (because of its next function adding 1 to the current ByteKey) this caused problems on the test case with a large number of keys. RedisCursorRangeTracker maintains the same logic used on ByteKeyRangeTracker with the exception of a modified trySplit method.

You can convert the next() ByteKey cursor by rounding up to next closet valid redis cursor. If the bytekey is 8 bytes (length of long) then you can decode it using the BigEndianLongCoder and if its longer then 8 bytes then you can truncate it down to 8 bytes decode it and add 1. This will allow you to reuse the ByteKeyRangeTracker.

Note that the ByteKeyRangeTracker is has a lot of logic which has been well tested so duplicating it is very likely to create subtle errors.

MiguelAnzoWizeline commented:
Hi, the problem I'm having with that kind of approach is that we cannot get a closest valid redis cursor, as we don't know what the next cursor is going to be until we send the actual cursor to redis. Obtaining a calculated cursor and sending it to redis would cause redis to return us a wrong cursor for the data we are trying to obtain.

lukecwik commented:
> 

You need to be able to use a generated Redis cursor otherwise your not going to have much value converting RedisIO to an SDF.

Try looking at https://engineering.q42.nl/redis-scan-cursor/ as it goes into a bunch of detail about how the cursors are translated back and forth. They even allude to that you can generate the cursors at arbitrary percentiles which is what we were looking for here. https://www.freecodecamp.org/news/redis-hash-table-scan-explained-537cc8bb9f52/ goes into the internal redis server implementation details.

MiguelAnzoWizeline commented:
Hi, checking again on the doc it should work correctly then using directly a ByteKey implementation, however in this implementation the test with a lot of records in redis is failing (testReadSplitBig which uses 1000 elements), it's always missing one random element in the response. I think it may be because of the splitting, I'm currently reviewing it in detail to see what it’s causing it.

MiguelAnzoWizeline commented:
Fixed the error that happened in the split test by removing the deprecated Reshuffle and adding a Distinct transform to remove duplicate results, all test are passing now.

aaltay commented:
What is the next step on this PR?

MiguelAnzoWizeline commented:
Hi @aaltay I’m waiting for a review from lukecwik, or if anyone else wants to give it a review also it would be great

aaltay commented:
@lukecwik - Could you take another look at this change please?

lukecwik commented:
@aaltay This is not a high priority for me. It would be best if someone else had the cycles.

I took a quick look and its great that we aren't copying the code for the restriction tracker.

One sticking point is that the PR still doesn't address the usage of dbSize() which isn't the same as table size and I haven't been thorough on the tests. For example, we should have a case where the table size is large and the dbSize is small. We should also verify that the results we get when reading is a superset of the original table when the table is being continuously added to and similarly we should start with a large table and then remove most of the items by starting with a large set of numbers and then remove everything except for numbers that divide by 7 ensuring that the numbers that are read match all the expected numbers that wouldn't have been removed.

MiguelAnzoWizeline commented:
Regarding the table size, what we need for the operations is the next power of two needed to represent the number of keys in the database, so we get the number of keys with dbSize() and then obtain the needed value with [this](https://github.com/apache/beam/blob/7f00a105e64494ced92a404ce16bf7fecd383172/sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java#L136) method.

lukecwik commented:
If you insert a million keys the dbsize will be a million and the table size will be ~2 million. Then you remove 999k keys and the dbsize is now 1k, is the table size now ~2k?

aaltay commented:
> @aaltay This is not a high priority for me. It would be best if someone else had the cycles.
> 
> I took a quick look and its great that we aren't copying the code for the restriction tracker.
> 
> One sticking point is that the PR still doesn't address the usage of dbSize() which isn't the same as table size and I haven't been thorough on the tests. For example, we should have a case where the table size is large and the dbSize is small. We should also verify that the results we get when reading is a superset of the original table when the table is being continuously added to and similarly we should start with a large table and then remove most of the items by starting with a large set of numbers and then remove everything except for numbers that divide by 7 ensuring that the numbers that are read match all the expected numbers that wouldn't have been removed.

> @aaltay This is not a high priority for me. It would be best if someone else had the cycles.

Ack. Will do.

aaltay commented:
@kileys - could you find another reviewer? Given the holiday timeline, you check with @ihji or you can do the review yourself. :)

kileys commented:
R: @pabloem 

Thanks Pablo!

aaltay commented:
@pabloem - would you be able to review this change?

aaltay commented:
@pabloem - friendly ping. Could you review or find another reviewer?

pabloem reviewed (commented):
ok looking at this now

pabloem commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -447,7 +388,7 @@ public void processElement(ProcessContext c) {
                       })
                   .withSideInputs(empty));
-      return materialized.apply(Reshuffle.viaRandomKey());
+      return materialized.apply(Distinct.create());
```
why are we using distinct here?

pabloem replied:
I am not ready to mark this as resolved yet - this approach using `Distinct` may be a problem in streaming, so I want to make sure I read the guarantees properly

pabloem replied:
Ok, let's update the documentation of the transform to indicate that redis may output duplicates, and so we have an extra step to deduplicate them. I think it makes sense to add an option `withAllowDuplicates` that increases performance but may have duplicate elements in a single PCollection

Also, instead of `Distinct`, let's use `Latest.perKey()` (https://beam.apache.org/documentation/transforms/java/aggregation/latest/)

pabloem replied:
@chamikaramj what are your thoughts on a parameter `withAllowDuplicates`? Or we could have the fewest possible knobs

fernando-wizeline replied:
About using withAllowDuplicates, should we proceed with the change or is there is something else to discuss about it?

MiguelAnzoWizeline commented:
Added a couple of comments to help keep a reference of why it is needed to remove duplicates from the results and how is the conversion between a Redis cursor and a ByteKey made.

pabloem commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java:
```diff
@@ -0,0 +1,161 @@
+    int nBits = getTablePow(cursor.getDbSize());
+    long cursorLong = Long.parseLong(cursor.getCursor());
+    long reversed = shiftBits(cursorLong, nBits);
+    BigEndianLongCoder coder = BigEndianLongCoder.of();
```
maybe create this coder only once. Perhaps add a member variable INSTANCE to the class BigEndianLongCoder, so that we will only create one instance, and have better performance

pabloem commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -334,95 +337,33 @@ public void setup() {
+            }
           }
         }
+        redisCursor = RedisCursor.of(scanResult.getCursor(), jedis.dbSize(), false);
```
it seems like we need to end this when `scanResult.getCursor()` returns 0 again. Is that correct? Is this case handled correctly?

fernando-wizeline replied:
Yeah, that is correct, we're handling the case properly.
This is being handled by the try/claim in line 353-361, since the ByteKey has a special case for the 0 value at the end of the cursor.

fernando-wizeline commented:
> @MiguelAnzoWizeline could you address @pabloem 's existing comments so that he can potentially continue this review ?

Hi @chamikaramj!
Miguel is no longer working on this project, but we will sync with him and provide a response in a couple of days.

fernando-wizeline commented:
Hi @pabloem!
How do the changes look?
Does it look good for merging?
Thanks for the feedback!

pabloem commented:
can you please fix merge conflicts, and make sure that all precommit tests are passing so that we can do the review? (see some related test failures: https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/21582/)

benWize commented:
Hi @pabloem, I'm working on the fixes for this PR, I found out that the new errors like https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/21582/testReport/junit/org.apache.beam.sdk.io.redis/RedisIOTest/testRead/ are related to the use of `Latest.perKey()`. 
The failing case for that combineFn happen when the elements in the PCollection don't contain timestamped values, so a timestamp is added in https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java#L188, and the condition in https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java#L110 return the accumulator with a null value, so when tries to encode produces the reported failure in the tests. I'm not sure if we need to add something else to RedisIO transform, or Latest combineFn should be modified to include that specific case.

benWize commented:
Run Java PreCommit

pabloem commented:
taking a look...

benWize commented:
Run Java PreCommit

aaltay commented:
What is the next step on this PR?

pabloem commented:
taking a look

pabloem commented:
ok this lgtm. can you please use Long.reverse in the comment and then we can merge? thanks @benWize

pabloem commented:
Run Java PostCommit

benWize commented:
Run Kotlin_Examples PreCommit

pabloem commented:
LGTM. Merging.

benWize commented:
Thank you, @pabloem, and all participants in this PR
